crates/turborepo-cache/src/cache_archive/create.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    backtrace::Backtrace,
+    fs,
+    fs::{File, OpenOptions},
+    io::{BufWriter, Read},
+    path::Path,
+};
+
+use tar::{EntryType, Header};
+use turbopath::{AbsoluteSystemPath, AnchoredSystemPath, RelativeUnixPathBuf};
+
+use crate::CacheError;
+
+// We use an enum to get around Rust's generic restrictions
+// i.e. you can't have a function that can return two different
+// versions of a generic type like Vec<u32> and Vec<u64>
+enum CacheArchive<'a> {
+    Compressed(tar::Builder<zstd::Encoder<'a, BufWriter<File>>>),
+    Uncompressed(tar::Builder<BufWriter<File>>),
+}
+
+impl<'a> CacheArchive<'a> {
+    // Appends data to tar builder.
+    fn append_data(
+        &mut self,
+        header: &mut Header,
+        path: impl AsRef<Path>,
+        body: impl Read,
+    ) -> Result<(), CacheError> {
+        match self {
+            CacheArchive::Compressed(builder) => Ok(builder.append_data(header, path, body)?),
+            CacheArchive::Uncompressed(builder) => Ok(builder.append_data(header, path, body)?),
+        }
+    }
+
+    // Makes a new CacheArchive at the specified path
+    // Wires up the chain of writers:
+    // tar::Builder -> zstd::Encoder (optional) -> BufWriter -> File
+    fn create(path: &AbsoluteSystemPath) -> Result<Self, CacheError> {
+        let mut options = OpenOptions::new();
+
+        let file = options
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path.as_path())?;
+
+        // Flush to disk in 1mb chunks.
+        let file_buffer = BufWriter::with_capacity(2usize.pow(20), file);
+
+        let is_compressed = path.as_path().ends_with(".zst");
+
+        if is_compressed {
+            let zw = zstd::Encoder::new(file_buffer, 0)?;
+
+            Ok(CacheArchive::Compressed(tar::Builder::new(zw)))
+        } else {
+            Ok(CacheArchive::Uncompressed(tar::Builder::new(file_buffer)))
+        }
+    }
+
+    // Adds a user-cached item to the tar
+    fn add_file(
+        &mut self,
+        anchor: &AbsoluteSystemPath,
+        file_path: &AnchoredSystemPath,
+    ) -> Result<(), CacheError> {
+        // Resolve the fully-qualified path to the file to read it.
+        let source_path = anchor.resolve(file_path);
+
+        // Grab the file info to construct the header.
+        let file_info = fs::symlink_metadata(source_path.as_path())?;
+
+        // Normalize the path within the cache
+        let file_path = RelativeUnixPathBuf::new(file_path.to_str()?.as_bytes())?;
+        let canonical_file_path = file_path.make_canonical_for_tar(file_info.is_dir());
```

Comment:
This one is probably a judgment call. The extra type is not _really_ buying us much here, since we immediately convert it to a `Path`. In this case it might be ok to have `file_path.to_canonical_tar_path()` or something like that return a `Path` directly. We're only doing this because the tar library's API requires a path, it then goes right back and does all of the work to turn it back into unix path bytes.